### ISSUE TITLE
[Perf -14%] System.Tests.Perf_UInt32.TryParseHex

### ISSUE THREAD

**Issue by DrewScoggins** [dotnet/runtime#39119]
### Run Information

Architecture | x64
-- | --
OS | ubuntu 18.04
Changes | [diff](https://github.com/dotnet/installer/compare/a4a49f07fbf986d496ef57521e76b22caaf9bfb3...3c6656f83b01ece6af7e11eb75ec0145554b139d)

### Regressions in System.Tests.Perf_UInt32

Benchmark | Baseline | Test | Test/Base | Modality | Baseline Outlier
-- | -- | -- | -- | -- | --
TryParseHex | 15.14 ns | 17.49 ns | 1.16 | Bimodal | False

![graph](<https://pvscmdupload.blob.core.windows.net/autofilereport/autofilereports/07_10_2020/release/3.1.1xx%20-%3E%20master_x64_ubuntu%2018.04/System.Tests.Perf_UInt32_1.png>)
[Historical Data in Reporting System](<https://pvscmdupload.blob.core.windows.net/autofilereport/autofilereports/07_10_2020/release/3.1.1xx%20-%3E%20master_x64_ubuntu%2018.04/System.Tests.Perf_UInt32.html>)

### Repro

```cmd
git clone https://github.com/dotnet/performance.git
py .\performance\scripts\benchmarks_ci.py -f netcoreapp5.0 --filter 'System.Tests.Perf_UInt32*';
```
<details>

#### System.Tests.Perf_UInt32.TryParseHex(value: "FFFFFFFF")

```log
[14.487 ; 14.918) | @@@
[14.918 ; 15.494) | @@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@
[15.494 ; 15.920) | @@@
[15.920 ; 16.496) | 
[16.496 ; 16.855) | 
[16.855 ; 17.250) | @@@@@@@@
[17.250 ; 17.826) | @@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@
[17.826 ; 18.411) | @@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@
[18.411 ; 18.967) | @@@@@@
[18.967 ; 19.543) | 
[19.543 ; 20.119) | 
[20.119 ; 20.695) | 
[20.695 ; 21.323) | 
[21.323 ; 21.899) | @
```

### Docs

[Profiling workflow for dotnet/runtime repository](https://github.com/dotnet/performance/blob/master/docs/profiling-workflow-dotnet-runtime.md)
[Benchmarking workflow for dotnet/runtime repository](https://github.com/dotnet/performance/blob/master/docs/benchmarking-workflow-dotnet-runtime.md)

</details>

**Comment by danmosemsft** [comment]
@DrewScoggins  please include a diff link showing changes closely bracketing regression. This one does not work. I believe this is important for all regression issues.

**Comment by DrewScoggins** [comment]
https://github.com/dotnet/runtime/issues/39122#issuecomment-656941297

**Comment by stephentoub** [comment]
I can't repro this locally on Ubuntu 18.04.  All inputs to this test show improvement for me:

```
dotnet run -c Release -f netcoreapp3.1 --runtimes netcoreapp3.1 netcoreapp5.0 --join --filter *System.Tests.Perf_UInt32.TryParseHex*
```

|      Method |        Job |       Runtime |     Toolchain |    value |      Mean |     Error |    StdDev |    Median |       Min |       Max | Ratio |
|------------ |----------- |-------------- |-------------- |--------- |----------:|----------:|----------:|----------:|----------:|----------:|------:|
| TryParseHex | Job-WVEAVH | .NET Core 3.1 | netcoreapp3.1 |        0 | 10.219 ns | 0.0542 ns | 0.0480 ns | 10.204 ns | 10.125 ns | 10.317 ns |  1.00 |
| TryParseHex | Job-ZMIWWJ | .NET Core 5.0 | netcoreapp5.0 |        0 |  8.613 ns | 0.0618 ns | 0.0578 ns |  8.583 ns |  8.554 ns |  8.732 ns |  0.84 |
|             |            |               |               |          |           |           |           |           |           |           |       |
| TryParseHex | Job-WVEAVH | .NET Core 3.1 | netcoreapp3.1 |     3039 | 14.377 ns | 0.1322 ns | 0.1104 ns | 14.356 ns | 14.227 ns | 14.556 ns |  1.00 |
| TryParseHex | Job-ZMIWWJ | .NET Core 5.0 | netcoreapp5.0 |     3039 | 13.265 ns | 0.0622 ns | 0.0552 ns | 13.271 ns | 13.148 ns | 13.355 ns |  0.92 |
|             |            |               |               |          |           |           |           |           |           |           |       |
| TryParseHex | Job-WVEAVH | .NET Core 3.1 | netcoreapp3.1 | FFFFFFFF | 19.740 ns | 0.1613 ns | 0.1347 ns | 19.765 ns | 19.464 ns | 19.950 ns |  1.00 |
| TryParseHex | Job-ZMIWWJ | .NET Core 5.0 | netcoreapp5.0 | FFFFFFFF | 18.601 ns | 0.1318 ns | 0.1233 ns | 18.604 ns | 18.430 ns | 18.842 ns |  0.94 |

**Comment by danmosemsft** [comment]
I looked at the diffs for each transition back and forth in Dec and Jan and nothing stands out as relevant. Since there's not a local repro, I don't think this issue is useful.

**Comment by DrewScoggins** [comment]
This commit,https://github.com/dotnet/runtime/commit/f6dc87f43df543bdc761ded7c578ecedc5b39668, is in the tight bound, https://github.com/dotnet/runtime/compare/49186d7b0e3d8ea0ae27a14cd3016e93311dda1f...b5705587347d29d79cec830dc22b389e1ad9a9e0. You can see on this [report](https://pvscmdupload.blob.core.windows.net/reports/allTestHistory%2frefs%2fheads%2fmaster_x64_ubuntu%2018.04%2fSystem.Tests.Perf_UInt32.TryParseHex(value%3a%20%22FFFFFFFF%22).html) that we are seeing a very noticeable jump. Given that the code change happened over that 12 hour period, it feels like it is worth some further digging. We can give access to a lab machine for repro as well if we are having trouble tracking it down on other machines.

**Comment by danmosemsft** [comment]
Thank you @DrewScoggins - that's from July 17, whereas the issue was for Nov and Mar regressions, and @stephentoub verified July 10th. Still we can reuse this issue.

We discussed a very similar regression due to this case here https://github.com/dotnet/runtime/issues/39720#issuecomment-669728539

The commit you correctly flagged was one that created a public Hex parsing API and replaced all other internal Hex parsing with this. As in that case, the API is slower for very short inputs. We decided to ignore that case because the absolute time is so tiny. In this case the absolute time is still tiny (regression is 2 nanoseconds) but it's an API that's more likely to be called often. 

@tannergooding owns this area and can make the best call on whether the code unification is worth the 2ns regression -- and in particular whether @tkp1n 's https://github.com/dotnet/runtime/pull/39702 would potentially regain this regression without losing that nice common implementation.
